url3f8d6440dd_huiwq1990_CognativeModel_tgz-pJ8-clarion_samples_HelloWorldJ8/src/clarion/system/HiddenRuleCollection.java

```diff
@@ -0,0 +1,436 @@
+	}
+	
+	/**
+	 * Initializes a rule collection with the collection of rules specified.
+	 * @param rules The rules for the collection.
+	 */
+	public HiddenRuleCollection (Collection <? extends AbstractRule> rules)
+	{
+		super();
+		for(AbstractRule r : rules)
+			add(r);
+	}
+	
+	/**
+	 * Gets a rule in the collection that matches the specified condition and 
+	 * action. If the collection does not contain a rule with the specified condition and 
+	 * action, this method returns null.
+	 * @param cond The condition of the rule to get.
+	 * @param act The action of the rules to get.
+	 * @return A rules with the specified condition and action.
+	 */
+	public AbstractRule get (GeneralizedConditionChunk cond, AbstractAction act)
+	{
+		HashMap<Object, HashMap<Object,LinkedList <AbstractRule>>> rc = get(act.getID());
+		if (rc == null)
+			return null;
+		if(cond == null)
+		{
+			//Only 1 condition-less rule is allowed for each action. This is only the case for a rule 
+			//that does not have a condition.
+			LinkedList <AbstractRule> candidates = rc.get(null).get(null);
+			if(candidates == null)
+				return null;
+			return candidates.iterator().next();
+		}
+		else
+		{
+			for(Dimension d : cond.values())
+			{
+				HashMap<Object,LinkedList <AbstractRule>> rrc = rc.get(d.getID());
+				if(rrc == null)
+					return null;
+				for(Value v : d.values())
+				{
+					if(v.isFullyActivated())
+					{
+						LinkedList <AbstractRule> candidates = rrc.get(v.getID());
+						if(candidates == null)
+							return null;
+						for(AbstractRule r : candidates)
+						{
+							if(r.getCondition().equals(cond))
+								return r;
+						}
+						return null;
+					}
+				}
+			}
+		}
+		//It is highly unlikely this return statement will be reached. The specified condition would have to 
+		//have no dimensions OR all of the values within all of the dimensions of the specified condition
+		//would have to be "off", in which case the condition would never match any state and would therefore
+		//not be very useful.
+		return null;
+	}
+	
+	/**
+	 * Checks to see if the collection contains the specified rule.
+	 * @param rule The rule object you wish to check for within this collection.
+	 * @return True if the collection contains the rule specified, otherwise false.
+	 */
+	public boolean contains (AbstractRule rule)
+	{
+		if(get(rule.rCondition, rule.rAction) != null)
+			return true;
+		return false;
+	}
+	
+	/**
+	 * Adds a rule into the rule collection. If the specified rule is equal to any rules 
+	 * currently in the collection, the rule will NOT be added. In addition, this method
+	 * also resets all of the rule variations, children, and match statistics from the specified 
+	 * rule (if the rule is refineable) since it is not necessary to track anything for rules that are 
+	 * not in the rule store.
+	 * @param R The rule to add.
+	 * @return True if the rule was successfully added, otherwise false.
+	 */
+	public boolean add (AbstractRule R)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.